tests/baselines/reference/api/typescript.d.ts

```diff
@@ -3955,19 +3955,42 @@ declare namespace ts {
+    /**
+     * A function that walks a node using the given visitor, lifting node arrays into single nodes,
+     * returning an node which satisfies the test.
+     *
+     * This type is complicated, but intends to encode the following behaviors:
+     *
+     *   - If the input node is potentially undefined, the output is potentially undefined.
+     *   - If the visitor can return undefined, the output is potentially undefined.
+     *   - If the output node is not undefined, then it will satisfy the test.
+     *
+     * @see {visitNode}
+     */
     export interface NodeVisitor {
-        <T extends Node>(nodes: T, visitor: Visitor | undefined, test?: (node: Node) => boolean, lift?: (node: readonly Node[]) => T): T;
-        <T extends Node>(nodes: T | undefined, visitor: Visitor | undefined, test?: (node: Node) => boolean, lift?: (node: readonly Node[]) => T): T | undefined;
+        <TIn extends Node | undefined, TVisited extends Node | undefined, TAssert extends NonNullable<TVisited>, TOut extends TIn extends undefined ? TAssert | undefined : TVisited extends undefined ? TAssert | undefined : TAssert>(node: TIn, visitor: Visitor<NonNullable<TIn>, TVisited> | undefined, test?: (node: Node) => node is TAssert, lift?: (node: readonly Node[]) => Node): TOut;
     }
+    /**
+     * A function that walks a node array using the given visitor, returning an array whose contents satisfy the test.
+     *
+     * This type is complicated, but intends to encode the following behaviors:
+     *
+     *   - If the input node array is potentially undefined, the output is potentially undefined.
+     *   - If the visitor can return undefined, the output may not be undefined; these nodes will be left in the output.
+     *   - If the output node array is not undefined, then its contents will satisfy the test.
+     *
+     * @see {visitNodes}
+     */
     export interface NodesVisitor {
-        <T extends Node>(nodes: NodeArray<T>, visitor: Visitor | undefined, test?: (node: Node) => boolean, start?: number, count?: number): NodeArray<T>;
-        <T extends Node>(nodes: NodeArray<T> | undefined, visitor: Visitor | undefined, test?: (node: Node) => boolean, start?: number, count?: number): NodeArray<T> | undefined;
+        <TIn extends Node, TArray extends NodeArray<TIn> | undefined, TVisited extends Node | undefined, TAssert extends NonNullable<TVisited>, TOutArray extends TArray extends undefined ? NodeArray<TAssert> | undefined : NodeArray<TAssert>>(nodes: TArray, visitor: Visitor<TIn, TVisited> | undefined, test?: (node: Node) => node is TAssert, start?: number, count?: number): TOutArray;
     }
-    export type VisitResult<T extends Node> = T | readonly T[] | undefined;
+    export function _jakeTesting(node: Node, nodeVisitor: Visitor<Node, Expression>): AssertionExpression;
+    export function _jakeTesting2(nodes: NodeArray<Node>, nodeVisitor: Visitor<Node, Expression | undefined>): NodeArray<AssertionExpression>;
```

Comment:
Ship it 😄

Reply:
Oh shoot, oops. Yeah, this is still a draft for a reason. This was me trying all of the variants to make sure that these types actually did the right thing.